gio-macros/src/actions.rs

```diff
@@ -0,0 +1,556 @@
+        } else {
+            quote_spanned! { state_type.span() =>
+                <#state_type as std::default::Default>::default()
+            }
+        };
+        quote! {
+            let initial_state: #state_type = #initial_state_expr;
+            let initial_state_variant = <#state_type as glib::variant::ToVariant>::to_variant(&initial_state);
+            let action = gio::SimpleAction::new_stateful(#action_name, #parameter, &initial_state_variant);
+        }
+    } else {
+        quote! {
+            let action = gio::SimpleAction::new(#action_name, #parameter);
+        }
+    };
+
+    let activate_handler = info
+        .activate_handler
+        .as_ref()
+        .map(|handler| generate_activate_handler(info, handler))
+        .transpose()?;
+
+    let change_state_handler = info
+        .change_state_handler
+        .as_ref()
+        .map(|handler| generate_change_state_handler(info, handler))
+        .transpose()?;
+
+    let action = quote! {
+        {
+            #create
+            #activate_handler
+            #change_state_handler
+            action
+        }
+    };
+    Ok(action)
+}
+
+fn combine_errors(error_acc: &mut Option<Error>, error: Error) {
+    match error_acc {
+        Some(ref mut error_acc) => {
+            error_acc.combine(error);
+        }
+        None => {
+            error_acc.replace(error);
+        }
+    }
+}
+
+fn attributes_to_metas(attributes: Vec<Attribute>) -> Result<Vec<NestedMeta>, Error> {
+    let mut metas = Vec::new();
+    let mut error = None;
+    for attr in attributes {
+        let meta = attr.parse_meta()?;
+        match meta {
+            Meta::List(MetaList { nested, .. }) => metas.extend(nested),
+            _ => combine_errors(&mut error, Error::new(attr.span(), "Unexpected attribute")),
+        }
+    }
+    if let Some(error) = error {
+        Err(error)
+    } else {
+        Ok(metas)
+    }
+}
+
+fn is_assoc(sig: &Signature) -> bool {
+    sig.inputs
+        .first()
+        .map_or(false, |arg| matches!(arg, FnArg::Receiver(..)))
+}
+
+fn argument_type(arg: &FnArg) -> Result<&Type, Error> {
+    match arg {
+        FnArg::Typed(PatType { ty, .. }) => Ok(&*ty),
+        _ => Err(Error::new(
+            arg.span(),
+            "Cannot extract type of an argument.",
+        )),
+    }
+}
+
+fn generate_register_method(
+    attrs: &ActionImplAttributes,
+    actions: &[TokenStream2],
+) -> ImplItemMethod {
+    let register_fn = format_ident!(
+        "{}",
+        attrs.register_fn.as_deref().unwrap_or("register_actions")
+    );
+    let register_method = quote! {
+        #[allow(clippy)]
+        fn #register_fn<AM: glib::object::IsA<gio::ActionMap>>(&self, map: &AM) {
+            #(
+                map.add_action(& #actions );
+            )*
+        }
+    };
+    parse(register_method.into()).unwrap()
+}
+
+pub fn actions(
+    attrs: ActionImplAttributes,
+    mut input: ItemImpl,
+) -> Result<TokenStream, TokenStream> {
```

Comment:
Instead of returning a `Result` and using `to_compile_error`, you can catch the error, pass it to `proc_macro_error::emit_error` and then generate an empty register method. I made the gtk4 macros do something like this and it works really well to reduce the error message if you have a nightly compiler.